File: examples/java/transaction/src/main/java/dev/angzarr/examples/transaction/TransactionState.java
package dev.angzarr.examples.transaction;

import examples.Domains.LineItem;
import java.util.List;

/**
 * Represents the current state of a transaction aggregate.
 */
public record TransactionState(
    String customerId,
    List<LineItem> items,
    int subtotalCents,
    int discountCents,
    String discountType,
    Status status
) {
    public enum Status {
        NEW, PENDING, COMPLETED, CANCELLED
    }

    public static TransactionState empty() {
        return new TransactionState("", List.of(), 0, 0, "", Status.NEW);
    }

    public boolean isPending() {
        return status == Status.PENDING;
    }

    public boolean isNew() {
        return status == Status.NEW;
    }

    public int calculateFinalTotal() {
        int total = subtotalCents - discountCents;
        return Math.max(0, total);
    }

    public int calculateLoyaltyPoints() {
        return calculateFinalTotal() / 100; // 1 point per dollar
    }
}

File: examples/kotlin/transaction/src/main/kotlin/dev/angzarr/examples/transaction/TransactionState.kt
package dev.angzarr.examples.transaction

import examples.Domains.LineItem

/**
 * Immutable transaction aggregate state.
 */
data class TransactionState(
    val customerId: String = "",
    val items: List<LineItem> = emptyList(),
    val subtotalCents: Int = 0,
    val discountCents: Int = 0,
    val discountType: String = "",
    val status: Status = Status.NONE
) {
    enum class Status {
        NONE,
        PENDING,
        COMPLETED,
        CANCELLED
    }

    fun exists(): Boolean = status != Status.NONE

    fun isPending(): Boolean = status == Status.PENDING

    fun finalTotal(): Int = subtotalCents - discountCents

    companion object {
        fun empty(): TransactionState = TransactionState()
    }
}
